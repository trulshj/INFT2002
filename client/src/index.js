// @flow

import ReactDOM from 'react-dom';
import * as React from 'react';
import { Component } from 'react-simplified';
import { HashRouter, Route } from 'react-router-dom';
import { NavBar, Card, Alert, Row, Column, Form, Button } from './widgets';
import Login from './components/login';
import NewQuiz from './components/newquiz';
import Quizzes from './components/quizzes';
import Register from './components/register';

class Menu extends Component {
  render() {
    return (
      <NavBar brand="Quiz App">
        <NavBar.Link to="/quizzes">Quizzes</NavBar.Link>
        <NavBar.Link to="/newQuiz">New Quiz</NavBar.Link>
        <NavBar.Link to="/login">Login</NavBar.Link>
      </NavBar>
    );
  }
}

class Home extends Component {
  render() {
    return <Card title="Welcome">Quiz App for Quizzes</Card>;
  }
}

const root = document.getElementById('root');
if (root)
  ReactDOM.render(
    <HashRouter>
      <div>
        <Alert />
        <Menu />
        <Route exact path="/" component={Home} />
        <Route exact path="/quizzes" component={Quizzes} />
        <Route exact path="/newQuiz" component={NewQuiz} />
        <Route exact path="/login" component={Login} />
        <Route exact path="/register" component={Register} />
      </div>
    </HashRouter>,
    root,
  );
