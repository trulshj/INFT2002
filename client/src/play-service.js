export type QuestionAnswers = {
  questionId: number,
  givenAnswers: boolean[],
  correctAnswers: boolean[],
};

class PlayService {
  questions: QuestionAnswers[] = [];

  addQuestionAnswers(answers) {
    this.questions.push(answers);
  }

  setAnswer(quiz_question_id, option_id, bool) {
    for (let q of this.questions) {
      if (q.questionId == quiz_question_id) {
        q.givenAnswers[option_id] = bool;
      }
    }
  }

  clearAnswers(): null {
    this.questions = [];
  }

  checkAnswers(): number {
    let points = 0;

    for (const question of this.questions) {
      console.table(question);
      if (question.givenAnswers.length == question.correctAnswers.length) {
        for (let i = 0; i < question.givenAnswers.length; i++) {
          if (question.givenAnswers[i] == true && question.correctAnswers[i] == true) {
            points++;
          } if (question.givenAnswers[i] == true && question.correctAnswers[i] == false) {
            points--;
          }
        }
      }
    }
    // Clear the answers, we don't need em
    this.clearAnswers();
    
    //console.log(points);
    return points;
  }
}

const playService = new PlayService();

export default playService;
