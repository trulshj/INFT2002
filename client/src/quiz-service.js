// @flow
import axios from 'axios';
axios.defaults.baseURL = 'http://localhost:3000/api/v1';

export type Quiz = {
  quizId: number,
  quizName: string,
  quizCategory: string,
  rating: number,
};

export type QuizQuestion = {
  quizQuestionId: number,
  quizId: number,
  question: string,
};

export type QuizQuestionOption = {
  quizQuestionOptionId: number,
  quizQuestionId: number,
  questionAnswer: string,
  isCorrect: boolean,
};

export type Category = {
  category_name: string,
};

export type QuestionDetails = {
  quizId: number,
  questionId: number,
  question: string,
  options: {
    optionId: number,
    questionOption: string,
    isCorrect: number,
  }[],
};

class QuizService {
  /**
   * Get quiz with given id.
   */
  get(quizId: number) {
    return axios.get<Quiz>('/quizzes/' + quizId).then((response) => response.data);
  }

  /**
   * Get all quizzes.
   */
  getAll() {
    return axios.get<Quiz[]>('/quizzes').then((response) => response.data);
  }

  /**
   * Get quizzes with given category
   */
  getQuizzesWithCategory(category: string) {
    return axios.get<Quiz[]>('/quizzes/category/' + category).then((response) => response.data);
  }
  /**
   * Getting quizzes with search
   */
  getQuizzesSearch(search: string) {
    return axios.get<Quiz[]>('quizzes/search/' + search).then((response) => response.data);
  }

  /**
   * Create new quiz having the given quizname.
   */
  create(quizName: string, quizCategory: string, user: string) {
    return axios
      .post<{}, { quizId: number }>('/quizzes', {
        quizname: quizName,
        category: quizCategory,
        user: user,
      })
      .then((response) => {
        //console.log(response);
        return response.data.quizid;
      });
  }

  /**
   * Delete quiz
   */
  deleteQuiz(quizId: number) {
    return axios.delete<void>('/quizzes/' + quizId).then((response) => response.data);
  }

  /**
   * Delete quiz questions
   * Needed for delete quiz and can't be the same as deleteQuestion as deleteQuestion uses quizQuestionId not quizId
   */
  deleteQuizQuestions(quizId: number) {
    return axios
      .delete<void>('/quizzes/' + quizId + '/questions')
      .then((response) => response.data);
  }

  /**
   * Update quiz
   */
  updateQuiz(quiz: Quiz) {
    return axios.put<Quiz, void>('/quizzes', quiz).then((response) => response.data);
  }

  /**
   * Get category having a given name
   */
  getCategory(category: string) {
    return axios.get<Category>('/categories/' + category).then((response) => response.data);
  }

  /**
   * Get all categories
   */
  getAllcategories() {
    return axios.get<Category[]>('/categories').then((response) => response.data);
  }

  /**
   * Get all details
   */
  getAllDetails() {
    return axios.get<QuestionAndOption[]>('/alldetails').then((response) => response.data);
  }

  /**
   * Get all questions
   */

  getAllquestions() {
    return axios.get<QuizQuestion[]>('/quizzes/questions').then((response) => response.data);
  }

  /**
   * Get all questions that belong to quiz with given quizId
   */
  getAllQuestionsInQuiz(quizId: number) {
    return axios
      .get<QuizQuestion[]>('/quizzes/' + quizId + '/questions')
      .then((response) => response.data);
  }

  /**
   * Get spesific question
   */
  getQuestion(quizQuestionId: number) {
    return axios
      .get<QuizQuestion>('quizzes/questions/' + quizQuestionId)
      .then((response) => response.data);
  }

  /**
   * Get question options for spesific question
   */
  getQuestionOption(quizQuestionId: number) {
    return axios
      .get<QuizQuestionOption[]>('/quizzes/questions/' + quizQuestionId + '/options')
      .then((response) => response.data);
  }

  /**
   * Get correct question option for spesific question
   */
  getQuestionOptionCorrect(quizQuestionId: number) {
    return axios
      .get<QuizQuestionOption[]>('/quizzes/questions/' + quizQuestionId + '/correct')
      .then((response) => response.data);
  }

  /**
   * Create new question having the given quizid
   */
  createQuestion(
    quizId: number,
    quizQuestion: string,
    option1: string,
    isCorrect1: boolean,
    option2: string,
    isCorrect2: boolean,
    option3: string,
    isCorrect3: boolean,
  ) {
    return axios
      .post<{}, { quizQuestionId: number }>('/newQuiz', {
        quizid: quizId,
        quizquestion: quizQuestion,
        option1: option1,
        iscorrect1: isCorrect1,
        option2: option2,
        iscorrect2: isCorrect2,
        option3: option3,
        iscorrect3: isCorrect3,
      })
      .then((response) => {
        //console.log(response);
        return response.data.quizquestionid;
      });
  }

  /**
   * Create new rating having the given quizid
   */
  createRating(rating: number, quizId: number) {
    return axios
      .post<{}, { rating: number }>('/quizzes/' + quizId  + "/" + rating)
      .then((response) => {
        //console.log(response);
        return response.data.rating;
      });
  }

  /**
   * Create new option to quizquestion having the given quizquestionid
   */
  createOption(
    quizQuestionId: number,
    questionAnswer1: string,
    isCorrect1: boolean,
    questionAnswer2: string,
    isCorrect2: boolean,
    questionAnswer3: string,
    isCorrect3: boolean,
  ) {
    return axios
      .post<{}, { quizQuestionOptionId: number }>('/newQuiz', {
        quizquestionid: quizQuestionId,
        questionanswer1: questionAnswer1,
        iscorrect1: isCorrect1,
        questionanswer2: questionAnswer1,
        iscorrect2: isCorrect1,
        questionanswer3: questionAnswer1,
        iscorrect3: isCorrect1,
      })
      .then((response) => {
        //console.log(response);
        return response.data.quizquestionoptionid;
      });
  }

  /**
   * Delete question with given question id
   */
  deleteQuestion(quizQuestionId: number) {
    return axios
      .delete<void>('/quizzes/question/' + quizQuestionId)
      .then((response) => response.data);
  }

  /**
   * Delete options with given question id
   */
  deleteOption(quizQuestionId: number) {
    return axios
      .delete<void>('/quizzes/question/' + quizQuestionId + '/options')
      .then((response) => response.data);
  }

  updateQuestion(question: QuizQuestion) {
    return axios
      .put<QuizQuestion, void>('/quizzes/questions', question)
      .then((response) => response.data);
  }

  updateOption(option: QuizQuestionOption) {
    return axios
      .put<QuizQuestionOption, void>('/quizzes/questions/options', option)
      .then((response) => response.data);
  }
}

const quizService = new QuizService();
export default quizService;
