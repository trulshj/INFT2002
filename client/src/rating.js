// @flow

import * as React from 'react';
import { Component } from 'react-simplified';
import ReactStars from "react-rating-stars-component";

export class StarRating extends Component {

    render(){
      const ratingChanged = (newRating) => {
        console.log(newRating);
      };
      return (
          <ReactStars
            count={5}
            onChange={ratingChanged}
            size={24}
            activeColor="#ffd700"
          />)
    }
  
  }