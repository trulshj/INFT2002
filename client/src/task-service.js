// @flow
import axios from 'axios';

axios.defaults.baseURL = 'http://localhost:3000/api/v2';

export type Task = {
  id: number,
  title: string,
  description: string,
  done: boolean,
};

class TaskService {
  /**
   * Get task with given id.
   */
  get(id: number) {
    return axios.get<Task>('/tasks/' + id).then((response) => response.data);
  }

  /**
   * Get all tasks.
   */
  getAll() {
    return axios.get<Task[]>('/tasks').then((response) => response.data);
  }

  /**
   * Create new task having the given title.
   *
   * Resolves the newly created task id.
   */
  create(title: string, description: string) {
    return axios
      .post<{}, { id: number }>('/tasks', { title: title, description: description })
      .then((response) => response.data.id);
  }

  /**
   * Update given task.
   */
  update(task: Task) {
    return axios.put<Task, void>('/tasks', task).then((response) => response.data);
  }

  /**
   * Delete task with given id.
   */
  delete(id: number) {
    return axios.delete<void>('/tasks/' + id).then((response) => response.data);
  }
}

const taskService = new TaskService();
export default taskService;
