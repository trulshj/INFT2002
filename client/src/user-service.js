// @flow
import axios from 'axios';
axios.defaults.baseURL = 'http://localhost:3000/api/v1';

export type User = {
  username: String,
  password: String,
};

class UserService {
  /**
   * Get quiz with given id.
   */
  login(username: string, password: string) {
    return axios
      .post<User, String>('/login', { username: username, password: password })
      .then((response) => console.log(response.data))
      .catch((error: Error) => {
        console.log(error);
      });
  }
  register(username: string, password: string) {
    return axios
      .post<User, String>('/register', { username: username, password: password })
      .then((response) => console.log(response.data))
      .catch((error: Error) => {
        console.log(error);
      });
  }
}

const userService = new UserService();
export default userService;
