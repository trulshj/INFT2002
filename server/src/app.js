// @flow

import express from 'express';
import taskRouter from './task-router';

/**
 * Express application.
 */
const app = express<express$Request, express$Response>();

app.use(express.json());

// Since API is not compatible with v1, API version is increased to v2
app.use('/api/v2', taskRouter);

export default app;
