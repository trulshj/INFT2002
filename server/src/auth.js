// @flow

const crypto = require('crypto');

// Returns salt of given length
function makeSalt(length: number) {
  if (length == null) {
    length = 12;
  } else if (typeof length !== 'number') {
    throw new Error('Length must be an integer');
  }
  return crypto.randomBytes(Math.ceil(length)).toString('hex').slice(0, length);
}

// Helper function that does the actualt hashing
function hasher(password: string, salt: string) {
  let hash = crypto.createHmac('sha512', salt);
  hash.update(password);

  let value = hash.digest('hex');
  return {
    salt: salt,
    hashedPassword: value,
  };
}

// Hashes a password with given salt
function hashPassword(password: string, salt: string) {
  if (password == null || salt == null) {
    throw new Error('Password and Salt must be provided');
  }
  if (typeof password !== 'string' || typeof salt !== 'string') {
    throw new Error('Password and Salt must be strings');
  }
  return hasher(password, salt);
}

export default { makeSalt, hashPassword };
