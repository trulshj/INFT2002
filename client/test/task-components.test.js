// @flow

import * as React from 'react';
import { TaskList, TaskNew, TaskDetails, TaskEdit } from '../src/task-components';
import { type Task } from '../src/task-service';
import { shallow } from 'enzyme';
import { Form, Button, Column } from '../src/widgets';
import { NavLink } from 'react-router-dom';

jest.mock('../src/task-service', () => {
  class TaskService {
    getAll() {
      return Promise.resolve([
        { id: 1, title: 'Les leksjon', description: 'Side 1', done: false },
        { id: 2, title: 'Møt opp på forelesning', description: 'Rom 1001', done: false },
        { id: 3, title: 'Gjør øving', description: 'Oppgave 6', done: false },
      ]);
    }

    create(title: string, description: string) {
      return Promise.resolve(4); // Same as: return new Promise((resolve) => resolve(4));
    }

    get(id: number) {
      return Promise.resolve({ id: 1, title: 'Les leksjon', description: 'Side 1', done: false });
    }

    update(task: Task) {
      return Promise.resolve();
    }

    delete(id: number) {
      return Promise.resolve();
    }
  }
  return new TaskService();
});

describe('Task component tests', () => {
  test('TaskList draws correctly', (done) => {
    const wrapper = shallow(<TaskList />);

    // Wait for events to complete
    setTimeout(() => {
      expect(
        wrapper.containsAllMatchingElements([
          <NavLink to="/tasks/1">Les leksjon</NavLink>,
          <NavLink to="/tasks/2">Møt opp på forelesning</NavLink>,
          <NavLink to="/tasks/3">Gjør øving</NavLink>,
        ])
      ).toEqual(true);
      done();
    });
  });

  test('TaskNew correctly sets location on create', (done) => {
    const wrapper = shallow(<TaskNew />);

    wrapper.find(Form.Input).simulate('change', { currentTarget: { value: 'Kaffepause' } });
    // $FlowExpectedError
    expect(wrapper.containsMatchingElement(<Form.Input value="Kaffepause" />)).toEqual(true);

    wrapper.find(Button.Success).simulate('click');
    // Wait for events to complete
    setTimeout(() => {
      expect(location.hash).toEqual('#/tasks/4');
      done();
    });
  });

  test('TaskDetails draws correctly with MatchingElement', (done) => {
    const wrapper = shallow(<TaskDetails match={{ params: { id: 1 } }} />);

    // Wait for events to complete
    setTimeout(() => {
      expect(wrapper.containsMatchingElement(<Column>Les leksjon</Column>)).toEqual(true);
      expect(wrapper.containsMatchingElement(<Column>Side 1</Column>)).toEqual(true);
      done();
    });
  });

  test('TaskDetails draws correctly with Snapshot', (done) => {
    const wrapper = shallow(<TaskDetails match={{ params: { id: 1 } }} />);

    // Wait for events to complete
    setTimeout(() => {
      expect(wrapper).toMatchSnapshot();
      done();
    });
  });

  test('TaskList sets TaskNew location correctly', (done) => {
    const wrapper = shallow(<TaskList />);

    wrapper.find(Button.Success).simulate('click');

    // Wait for events to complete
    setTimeout(() => {
      expect(location.hash).toEqual('#/tasks/new');
      done();
    });
  });

  test('TaskDetails sets TaskEdit location correctly', (done) => {
    const wrapper = shallow(<TaskDetails match={{ params: { id: 1 } }} />);

    wrapper.find(Button.Success).simulate('click');

    // Wait for events to complete
    setTimeout(() => {
      expect(location.hash).toEqual('#/tasks/1/edit');
      done();
    });
  });

  test('TaskEdit draws correctly', (done) => {
    const wrapper = shallow(<TaskEdit match={{ params: { id: 1 } }} />);

    wrapper.find(Form.Input).simulate('change', { currentTarget: { value: 'test' } });

    // Wait for events to complete
    setTimeout(() => {
      // $FlowExpectedError
      expect(wrapper.containsMatchingElement(<Form.Input value="test" />));
      done();
    });
  });
});
