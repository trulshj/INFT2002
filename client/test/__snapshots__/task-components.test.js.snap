// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task component tests TaskDetails draws correctly with Snapshot 1`] = `
<Fragment>
  <Card
    title="Task"
  >
    <Row>
      <Column
        width={2}
      >
        Title:
      </Column>
      <Column>
        Les leksjon
      </Column>
    </Row>
    <Row>
      <Column
        width={2}
      >
        Description:
      </Column>
      <Column>
        Side 1
      </Column>
    </Row>
    <Row>
      <Column
        width={2}
      >
        Done:
      </Column>
      <Column>
        <FormCheckbox
          checked={false}
          disabled={true}
          onChange={[Function]}
        />
      </Column>
    </Row>
  </Card>
  <ButtonSuccess
    onClick={[Function]}
  >
    Edit
  </ButtonSuccess>
</Fragment>
`;
